eth2/utils/eth2_key_derivation/src/derived_key.rs

```diff
@@ -0,0 +1,822 @@
+///
+/// Indirectly defined in EIP-2333.
+pub const LAMPORT_ARRAY_SIZE: u8 = 255;
+
+/// The order of the BLS 12-381 curve.
+///
+/// Defined in EIP-2333.
+pub const R: &str = "52435875175126190479447740508185965837690552500527637822603658699938581184513";
+
+/// The `L` value used in the `hdkf_mod_r` function.
+///
+/// In EIP-2333 this value is defined as:
+///
+/// `ceil((1.5 * ceil(log2(r))) / 8)`
+pub const MOD_R_L: usize = 48;
+
+/// A BLS secret key that is derived from some `seed`, or generated as a child from some other
+/// `DerivedKey`.
+///
+/// Implements `Zeroize` on `Drop`.
+// It's not strictly necessary that `DerivedKey` implements `Zeroize`, but it seems prudent to be a
+// little over-cautious here; we don't require high-speed key generation at this stage.
+#[derive(Zeroize)]
+#[zeroize(drop)]
+pub struct DerivedKey(SecretHash);
+
+impl DerivedKey {
+    /// Instantiates `Self` from some secret seed bytes.
+    ///
+    /// The key is generated deterministically; the same `seed` will always return the same `Self`.
+    ///
+    /// ## Errors
+    ///
+    /// Returns `Err(())` if `seed.is_empty()`, otherwise always returns `Ok(self)`.
+    pub fn from_seed(seed: &[u8]) -> Result<Self, ()> {
+        if seed.is_empty() {
+            Err(())
+        } else {
+            Ok(Self(derive_master_sk(seed)))
+        }
+    }
+
+    /// Derives a child key from the secret `Self` at some `index`.
+    pub fn child(&self, index: u32) -> DerivedKey {
+        Self(derive_child_sk(self.0.as_bytes(), index))
+    }
+
+    /// Returns the secret BLS key in `self`.
+    pub fn secret(&self) -> &[u8] {
+        self.0.as_bytes()
+    }
+}
+
+/// Derives the "master" BLS secret key from some `seed` bytes.
+///
+/// Equivalent to `derive_master_SK` in EIP-2333.
+fn derive_master_sk(seed: &[u8]) -> SecretHash {
+    hkdf_mod_r(seed)
+}
+
+/// From the given `parent_sk`, derives a child key at index`.
+///
+/// Equivalent to `derive_child_SK` in EIP-2333.
+fn derive_child_sk(parent_sk: &[u8], index: u32) -> SecretHash {
+    let compressed_lamport_pk = parent_sk_to_lamport_pk(parent_sk, index);
+    hkdf_mod_r(compressed_lamport_pk.as_bytes())
+}
+
+/// From the `ikm` (initial key material), performs a HKDF-Extract and HKDF-Expand to generate a
+/// BLS private key within the order of the BLS-381 curve.
+///
+/// Equivalent to `HKDF_mod_r` in EIP-2333.
+fn hkdf_mod_r(ikm: &[u8]) -> SecretHash {
+    let prk = hkdf_extract("BLS-SIG-KEYGEN-SALT-".as_bytes(), ikm);
+    let okm = &hkdf_expand(prk.as_bytes(), MOD_R_L);
+    mod_r(okm.as_bytes())
+}
+
+/// Interprets `bytes` as a big-endian integer and returns that integer modulo the order of the
+/// BLS-381 curve.
+///
+/// This function is a part of the `HKDF_mod_r` function in EIP-2333.
+fn mod_r(bytes: &[u8]) -> SecretHash {
+    let n = BigUint::from_bytes_be(bytes);
+    let r = BigUint::parse_bytes(R.as_bytes(), 10).expect("must be able to parse R");
+    let x = (n % r).to_bytes_be();
```

Comment:
`n` and `x` as `BigUint` are not being zeroized. Do you think it is worth handling these?

Reply:
Great pickup! I've ensure I wrap the bytes in `SecretBytes` and swapped over to [`num-bigint-dig`](https://crates.io/crates/num-bigint-dig) to ensure we can zeroize those bigints.